### ISSUE TITLE
Segfault at libqpdf/QPDF.cc:131

### ISSUE THREAD

**Issue by zdohnal** [qpdf/qpdf#140]
Segfault after applying #51, #99, #100 and #101 patches to version 6.0.0, which happens in QPDF::processFile() method:

[New LWP 4226]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `MP640-TurboPrint 21 dpotter TNA_R39_0284_0284H_010.jpg 1 zedoDuplexadvanced=aut'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  _int_malloc (av=av@entry=0x7f7669981ae0 <main_arena>, bytes=bytes@entry=64) at malloc.c:3517
3517	          bck->fd = unsorted_chunks (av);

Thread 1 (Thread 0x7f766c1ba8c0 (LWP 4226)):
#0  _int_malloc (av=av@entry=0x7f7669981ae0 <main_arena>, bytes=bytes@entry=64) at malloc.c:3517
        iters = <optimized out>
        nb = <optimized out>
        idx = 5
        bin = <optimized out>
        victim = 0x55f6a9a95000
        size = 4608
        victim_index = <optimized out>
        remainder = <optimized out>
        remainder_size = <optimized out>
        block = <optimized out>
        bit = <optimized out>
        map = <optimized out>
        fwd = <optimized out>
        bck = 0x0
        errstr = 0x0
        __func__ = "_int_malloc"
#1  0x00007f76696429b8 in __GI___libc_malloc (bytes=bytes@entry=64) at malloc.c:2890
        ar_ptr = 0x7f7669981ae0 <main_arena>
        victim = <optimized out>
        hook = <optimized out>
#2  0x00007f7669f44418 in operator new (sz=sz@entry=64) at ../../../../libstdc++-v3/libsupc++/new_op.cc:50
        p = <optimized out>
#3  0x00007f766bdcdffd in QPDF::processFile (this=this@entry=0x55f6a9a94a20, filename=filename@entry=0x7ffd7f298c1d "/var/spool/cups/d00021-001", password=password@entry=0x0) at libqpdf/QPDF.cc:131
        fi = <optimized out>
#4  0x000055f6a890dcd8 in QPDF_PDFTOPDF_Processor::loadFilename (this=0x55f6a9a93a30, name=0x7ffd7f298c1d "/var/spool/cups/d00021-001") at filter/pdftopdf/qpdf_pdftopdf_processor.cc:427
No locals.
#5  0x000055f6a89045f4 in main (argc=7, argv=0x7ffd7f296a18) at filter/pdftopdf/pdftopdf.cc:928
        param = {jobId = 21, numCopies = 1, user = 0x7ffd7f2988a6 "dpotter", title = 0x7ffd7f2988ae "TNA_R39_0284_0284H_010.jpg", fitplot = false, page = {top = 833.400024, left = 9.64999962, right = 585.650024, bottom = 14.3999996, width = 595.299988, height = 841.900024}, orientation = ROT_0, normal_landscape = ROT_270, paper_is_landscape = false, duplex = true, border = NONE, nup = {nupX = 1, nupY = 1, width = nan(0x400000), height = nan(0x400000), landscape = false, first = X, xstart = LEFT, ystart = RIGHT, xalign = CENTER, yalign = CENTER}, reverse = false, pageLabel = "", evenPages = true, oddPages = true, pageRange = {static npos = 2147483647, data = std::vector of length 1, capacity 1 = {{first = 1, second = 2147483647}}}, mirror = false, xpos = CENTER, ypos = CENTER, collate = false, evenDuplex = false, booklet = BOOKLET_OFF, bookSignature = -1, autoRotate = true, emitJCL = true, deviceCopies = 1, deviceReverse = false, deviceCollate = false, setDuplex = false, page_logging = 0, copies_to_be_logged = 1}
        num_options = <optimized out>
        proc = std::unique_ptr<PDFTOPDF_Processor> containing 0x55f6a9a93a30
        options = 0x55f6a9a93a50
        ppd = 0x55f6a9a93540

**Comment by jberkenbilt** [comment]
Can you supply a file to reproduce this?

**Comment by jberkenbilt** [comment]
@zdohnal  Thanks for the segfault bug reports. Please supply sample files so I can reproduce this. How did you apply the patches? I didn't test just applying those patches, so I can't guarantee that's going to work. Some of the patches introduce binary incompatibility with 6.0.0, but I have not yet bumped the soname on the library. If you can provide sample files to reproduce the problems, I will test on master. If there's still a problem, I'll fix. Otherwise, this is probably something about how you're applying the patches. Thanks!

**Comment by zdohnal** [comment]
@jberkenbilt Oh, I didn't know that some of these patches causes binary incompatibility (it is because of new class, if I think about it now. Am I correct?), so all segfaults can be caused by 'not-rebuilding' cups-filters with new qpdf update. I'll try it and let you know, if it really was the case. For now you can ignore all segfault issues reported by me.

**Comment by jberkenbilt** [comment]
The binary incompatibility was caused by adding new private member variables inside QPDF. If I knew now what I new when I originally designed this 15 years ago, I would have used a level of indirection to avoid this problem. Since I had to break ABI compatibility anyway, I also took the opportunity to make a few breaking changes to other places in the interface, though I have maintained source compatibility.
